### PR TITLE
Models

### PR DESCRIPTION
**Pull request by shechframgia**
This is for Models (Ticket No. 55685)

### CODE REVIEW

File: app/Http/Controllers/Auth/AuthController.php [hunk 1]
```diff
@@ -67,6 +68,7 @@ protected function create(array $data)
             'name' => $data['name'],
             'email' => $data['email'],
             'password' => bcrypt($data['password']),
+            'role' => '2',
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/Auth/AuthController.php]: Don't use fixed value in your code, please define constants in your `app/Model/User.php`:
```
const TYPE_ADMIN = 1;
const TYPE_MEMBER = 2;
```
And you can user `\App\Models\User::TYPE_ADMIN` where you want to place.

File: app/Models/Activity.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+class Activity extends Model
+{
+    public function lessons()
+    {
+        return $this->hasMany(Lesson::class);
+    }
+
+    public function users()
+    {
+        return $this->belongsToMany(User::class);
```
Comment (nguyenthanhtung88) [inline app/Models/Activity.php]: Do you think it won't work? with `belongsToMany` relationship, we need `pivot` table `user_activities` that i can't see in your DB Schema.

File: app/Models/Lesson.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    public function words()
+    {
+        return $this->hasMany(Lesson::class);
+    }
+
+    public function activities()
+    {
+        return $this->belongsToMany(Activity::class);
```
Comment (nguyenthanhtung88) [inline app/Models/Lesson.php]: Same problem with https://github.com/framgia/fels_119/pull/5/files#r55020013

File: app/Models/Lesson_Words.php [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Model;
+
+class Lesson_Words extends Model
+{
+    //
```
Comment (nguyenthanhtung88) [inline app/Models/Lesson_Words.php]: Remove this line

File: app/Models/User.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    public function lessons()
+    {
+        return $this->hasMany(Lesson::class);
+    }
+
+    public function followers()
+    {
+        return $this->belongsToMany(Follow::class);
```
Comment (nguyenthanhtung88) [inline app/Models/User.php]: It should be `hasMany`

File: app/Models/User.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    public function followings()
+    {
+        return $this->belongsToMany(Follow::class);
```
Comment (nguyenthanhtung88) [inline app/Models/User.php]: It should be `hasMany`

File: app/Models/Lesson_Words.php [hunk 7]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Model;
+
+class Lesson_Words extends Model
```
Comment (nguyenthanhtung88) [inline app/Models/Lesson_Words.php]: file name should be `LessonWord.php` and class name should be `LessonWord`

File: app/Models/User_Words.php [hunk 8]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Model;
+
+class User_Words extends Model
+{
+    //
```
Comment (nguyenthanhtung88) [inline app/Models/User_Words.php]: Remove unnecessary line

File: app/Models/User_Words.php [hunk 9]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Model;
+
+class User_Words extends Model
```
Comment (nguyenthanhtung88) [inline app/Models/User_Words.php]: File name should be `UserWord.php` and class name should be `UserWord`

File: app/Models/Lesson.php [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Model;
+
+class Lesson extends Model
+{
+    public function users()
+    {
+        return $this->belongsToMany(User::class);
```
Comment (nguyenthanhtung88) [inline app/Models/Lesson.php]: https://github.com/framgia/fels_119/pull/5/files#r55020013

File: app/Models/User.php [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+     *
+     * @var array
+     */
+    protected $hidden = [
+        'password', 'remember_token',
+    ];
+
+    public function words()
+    {
+        return $this->belongsToMany(Word::class);
```
Comment (nguyenthanhtung88) [inline app/Models/User.php]: Did you read https://laravel.com/docs/5.1/eloquent-relationships#many-to-many ?
If you didn't define pivot table and foreign key here, you have to create `user_word` table (not `user_words` you designed)
Or with your DB Schema you have to define this relationship like:
`$this->belongsToMany(Word::class, 'user_words', 'user_id', 'word_id')`

I saw all `belongsToMany` relationship in your PR have same probelm, please refactor this issue :(

File: app/Models/Activity.php [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Model;
+
+class Activity extends Model
+{
+    public function lessons()
+    {
+        return $this->hasMany(Lesson::class);
```
Comment (nguyenthanhtung88) [inline app/Models/Activity.php]: I think this relationship must be ``belongsTo`.
`Activity` belongsTo `Lesson` and `Lesson` hasMany `Activity` right ?

File: app/Models/Lesson.php [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    public function words()
+    {
+        return $this->hasMany(Lesson::class);
+    }
+
+    public function activities()
+    {
+        return $this->belongsTo(Activity::class);
```
Comment (nguyenthanhtung88) [inline app/Models/Lesson.php]: It should be `hasMany`

File: app/Models/Lesson.php [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    public function category()
+    {
+        return $this->belongsTo(Category::class);
+    }
+
+    public function words()
+    {
+        return $this->hasMany(Lesson::class);
```
Comment (nguyenthanhtung88) [inline app/Models/Lesson.php]: Lesson `hasMany(Lesson::class)` ???

File: app/Models/Lesson.php [hunk 15]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Model;
+
+class Lesson extends Model
+{
+    public function users()
```
Comment (nguyenthanhtung88) [inline app/Models/Lesson.php]: Because of using `belongsTo` relationship, i suggest your function name should be singular (ex: `user` instead of `users` -> this is used for `hasMany`)

File: app/Models/Activity.php [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+    }
+
+    public function users()
```
Comment (nguyenthanhtung88) [inline app/Models/Activity.php]: Function name should be singular
https://github.com/framgia/fels_119/pull/5/files#r55159341

File: app/Models/Word.php [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+    use SoftDeletes;
+
+    public function category()
+    {
+        return $this->belongsTo(Category::class);
+    }
+
+    public function users()
+    {
+        return $this->belongsToMany(Lesson::class);
```
Comment (nguyenthanhtung88) [inline app/Models/Word.php]: - function name: `users`
- relationship: `belongsToMany(Lesson::class)`
What did you mean?

File: database/migrations/2014_10_12_000000_create_users_table.php [hunk 18]
```diff
@@ -1,37 +0,0 @@
-<?php
```
Comment (nguyenthanhtung88) [inline database/migrations/2014_10_12_000000_create_users_table.php]: Why did you make deleting this migration file ???

Comment (shechframgia) [inline database/migrations/2014_10_12_000000_create_users_table.php]: I made a mistake but I have a backup for it.

File: app/Models/Lesson.php [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    public function category()
+    {
+        return $this->belongsTo(Category::class);
+    }
+
+    public function words()
+    {
+        return $this->hasMany(Word::class);
```
Comment (nguyenthanhtung88) [inline app/Models/Lesson.php]: `hasMany` or `belongsToMany` ???

File: app/Models/Word.php [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+    use SoftDeletes;
+
+    public function category()
+    {
+        return $this->belongsTo(Category::class);
+    }
+
+    public function users()
+    {
+        return $this->belongsToMany(User::class);
```
Comment (nguyenthanhtung88) [inline app/Models/Word.php]: You have to define full detail of `belongsToMany` relationship, because your pivot table `user_words` didn't follow table naming convention.

File: app/Models/Word.php [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    public function lessons()
+    {
+        return $this->belongsToMany(Lesson::class);
```
Comment (nguyenthanhtung88) [inline app/Models/Word.php]: Same problem with https://github.com/framgia/fels_119/pull/5/files#r55170145

File: database/migrations/2014_10_12_000000_create_users_table.php (hunk 18 above)
Comment (nguyenthanhtung88) [inline database/migrations/2014_10_12_000000_create_users_table.php]: @shechframgia: You have to revert this change, your backup file still working with your system but i think if you're working with other members, it will raise problem.

**Comment by nguyenthanhtung88** [comment]
OK
